python/PyQt6/core/__init__.py.in

```diff
@@ -591,29 +591,58 @@ try:
    QgsRasterLayer.as_numpy = _raster_layer_as_numpy
 
    def _qgsgeometry_as_numpy(self) -> _typing.Union[_numpy.ndarray, _typing.List[_numpy.ndarray]]:
+       wkb = self.wkbType()
+       hasM = QgsWkbTypes.hasM(wkb)
+       hasZ = QgsWkbTypes.hasZ(wkb)
+
+       def get_xyzm_coordinates(pt):
+                 if hasZ and hasM:
+                     return _numpy.array([pt.x(), pt.y(), pt.z(), pt.m()])
+                 elif hasZ:
+                     return _numpy.array([pt.x(), pt.y(), pt.z()])
+                 elif hasM:
+                     return _numpy.array([pt.x(), pt.y(), pt.m()])
+                 else:
+                     return _numpy.array([pt.x(), pt.y()])
+
+       def fill_structure_with_elements(lst: _typing.List, elements: _typing.List, idx: int=0):
+           for i in range(len(lst)):
+               if isinstance(lst[i], list):
+                    idx = fill_structure_with_elements(lst[i], elements, idx)
+               else:
+                   lst[i] = _numpy.array(elements[idx])
+                   idx += 1
+           return idx
+
        if self.isMultipart():
-           if self.type() == QgsWkbTypes.PointGeometry:
-               # MultiPoint
-               return [_numpy.array([pt.x(), pt.y()]) for pt in self.asMultiPoint()]
-           elif self.type() == QgsWkbTypes.LineGeometry:
-               # MultiLineString
-               return [_numpy.array([[pt.x(), pt.y()] for pt in line]) for line in self.asMultiPolyline()]
-           elif self.type() == QgsWkbTypes.PolygonGeometry:
-               # MultiPolygon
-               return [
-                   [_numpy.array([[pt.x(), pt.y()] for pt in ring]) for ring in polygon]
-                   for polygon in self.asMultiPolygon()
-               ]
+            geometry_type = self.type()
```

Comment:
that variable can be moved before line 617, and used in the else block of lines 636 to 645